AIP-39: ``DagRun.data_interval_start`` and ``data_interval_end``

uranusjr opened the pull request:
Close #16304. Todo:

* [x] Add those fields to the actual `DagRun` model.
* [x] Populate the fields when a `DagRun` instance is created (when `DagRun.execution_date` is populated). I think (hope) this field is never updated after a run is created?
* [x] Add `logical_date` and `data_interval` to template context.
    * [x] Emit deprecation warning when the `execution_date` context variable is accessed. (fix #16372)
* Change functions that create `DagRun` instances to use `data_interval`.
    * [x] `DAG.create_dagrun()`
    * [x] Connexion `post_dag_run()`
    * [ ] (Optional) A lot of tests
* Change functions that call `DAG.create_dagrun()` to use `data_interval`.
	* [x] `BackfillJob`
	* [x] `SchedulerJob`
	* [x] `SubDagOperator`
	* [x] Web UI manual trigger view
	* [ ] (Optional) A lot of tests
	* [ ] (Optional) Experiemental API
* [x] Add documentation on how the template field meanings change and how they map to the previous values.
* Test data interval is correctly populated when triggered
	* [x] In the scheduler (manual, scheduled, and backfilled, etc.)
	* [x] From the web UI
	* [x] From the CLI
	* From the stable API (this needs things in #17122 and will be a part of that PR instead)
* Add documentation on data intervals. (Broken into another PR since I think there are many things to change)

uranusjr committed:
Implement DagRun.data_interval

This creates two new columns, data_interval_start and data_interval_end,
on DagRun, and corresponding fields in DagModel to forward them to
DagRun. Code initialiting a DagRun, including DAG.create_dagrun() and
its callers, are modified to calculate the run's data interval and pass
it to DagRun.

The data interval of a DAG run is as defined by its timetable, and can
be None if the timetable does not have a "schedule", e.g. DAGs with
schedule_interval set to None or "@once". Manual runs of a schedule have
their data intervals set to the previous period that would have been
scheduled if the scheduler triggers a scheduled run for that point of
time, e.g. for a DAG run on every midnight, a manually triggered run has
data interval of the previous day (midnight to midnight).

DAG runs created prior to this path have their data interval all set to
None, regardless of their schedule interval.

Migration for DagRun.date_interval_[start|end]

Migrate DAG.create_dagrun() to use data_interval

Fix data interval shim for non-schedule timetables

Fallback for tests without DAG serialization

Use built-in session management in migrations

Use mssql-specific column type in migration

Give up on backfilling columns

Use new dagrun_info format when populating DagModel

Pass data_interval around when scheduling stuff

Jump through hoops to make nametuple subclass work

ashb commented on airflow/migrations/versions/142555e44c17_add_data_interval_start_end_to_dagrun.py:
```diff
@@ -0,0 +1,100 @@
+import alembic
+import sqlalchemy
+
+from airflow.models import DagModel, DagRun
```
Rather than importing the model here, we should create a minimal version here -- otherwise this will _always_ use the latest version of the model meaning this might fail in the future.

See `airflow/migrations/versions/6e96a59344a4_make_taskinstance_pool_not_nullable.py` for example

uranusjr replied:
I chose to only pull out a minimal set of fields we need in the migration (a trick I sometimes do to make the migrations more readable).

ashb commented on airflow/migrations/versions/142555e44c17_add_data_interval_start_end_to_dagrun.py:
```diff
@@ -0,0 +1,100 @@
+        dag = serialized.dag
+        data_interval_start = dag_run.execution_date
+        dag_run.data_interval_start = data_interval_start
+        dag_run.data_interval_end = dag.following_schedule(data_interval_start)
```
Is this at all possible without requiring the seriazlied dag table? -- inflating every dag, and walking over every dag run is not going to be quick.

There is _a_ `schedule_interval` column on DagModel currently, that I _think_ gives us enough info? Not sure it saves us much run time though.

uranusjr replied:
Hmm, I just tried that, but… 

https://github.com/apache/airflow/blob/6e9e56246b216a43eabb050c5b220f3665de6305/airflow/models/dag.py#L481-L491

When calculating the data interval’s end (i.e. `next_execution_date`), we need to “fix” DST, which needs to know what timezone the DAG is in. But that information is only available in `DAG` (by inspecting the original `DAG.start_date` input) 😢

ashb replied:
Dang.

ashb commented on airflow/migrations/versions/142555e44c17_add_data_interval_start_end_to_dagrun.py:
```diff
@@ -0,0 +1,100 @@
+        DagRun.data_interval_end: DagRun.execution_date,
+    }
+    joined = session.query(DagRun).join(DagModel, DagModel.dag_id == DagRun.dag_id)
+    joined.filter(NO_SCHEDULE_FILTER).update(updates)
```
Add `synchronize_session=False` -- we don't need to update objects in the session here.

ashb commented on airflow/migrations/versions/142555e44c17_add_data_interval_start_end_to_dagrun.py:
```diff
@@ -0,0 +1,155 @@
+    """Apply add data_interval_[start|end] to DagRun."""
+    # Create aolumns with NULL as default.
+    with op.batch_alter_table("dag_run") as batch_op:
+        batch_op.add_column(Column("data_interval_start", UtcDateTime))
```
Is this the right type for mssql? Paging @aneesh-joseph

uranusjr replied:
Hmm CI for MSSQL is stuck during this migration so I likely am doing things wrong.

ephraimbuddy replied:
MSSQL needs `mssql.DATETIME2(precision=6)`  for the column

ashb commented on airflow/models/dag.py:
```diff
@@ -1812,19 +1815,40 @@ def create_dagrun(
+            if run_type == DagRunType.MANUAL:
+                data_interval = (start, start)
+            else:
+                data_interval = (start, self.following_schedule(start))
```
Shouldn't this go directly to a method on `self.timetable`?

uranusjr replied:
It should, but that’s not in `main` yet 🙂

uranusjr replied:
Now it is, so I’ll change this.

ashb commented on airflow/models/dag.py:
```diff
@@ -1812,19 +1815,40 @@ def create_dagrun(
         run = DagRun(
             dag_id=self.dag_id,
             run_id=run_id,
-            execution_date=execution_date,
```
The AIP calls for a column called `schedule_date` https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-39+Richer+scheduler_interval#AIP39Richerscheduler_interval-RenameDagRun.execution_datetoschedule_date

I'm happy to change the name of that column, but I think still we need _a_ canonical date on DagRun table to enable chronological sorting, don't we?

uranusjr replied:
We should, I just haven’t reached that part yet.

ashb commented on airflow/models/dagrun.py:
```diff
@@ -123,10 +127,47 @@ def __init__(
+
+                with create_session() as session:
+                    serialized = session.query(SerializedDagModel)
+                    dag = serialized.filter(SerializedDagModel.dag_id == dag_id).first().dag
```
```suggestion
                    dag = serialized.filter(SerializedDagModel.dag_id == dag_id).one().dag
```

uranusjr committed:
Add template context for AIP-39 variables

The following fields are deprecated:

* execution_date (renamed to schedule_date)
* next_execution_date (renamed to next_schedule_date)
* prev_execution_date (see below)
* prev_execution_date_success (renamed to prev_schedule_date_success)
* prev_ds
* prev_ds_nodash
* tomorrow_ds
* tomorrow_ds_nodash
* yesterday_ds
* yesterday_ds_nodash

A new field prev_schedule_date is added to replacement the functionality
provided by prev_execution_date, but with a different implementation.
The old 'execution' one is calculated in abstract by substracting an
interval from execution_date/schedule_date, but the new 'schedule' one
is based on the *actual* previous DagRun recorded in the database. The
new behavior matches prev_execution_date_success and is considered a
better implementation, but introduces incompatibility in certain edge
cases, such as when a DAG is first run (ever, or after a pause/unpause).

The following new fields are added for the new data interval concept
introduced in AIP-39:

* data_interval_start
* data_interval_end

This commit also includes a minor refactor on how a TaskInstance without
an associated DagRun is handled. This is impossible in practice, but is
depended on by many tests in Airflow's main test suite.

uranusjr committed:
Propogate data interval from context to subdag

uranusjr committed:
Inject lazy-object-proxy into virtual environment

This is needed to deepcopy compatibility context variables.

uranusjr committed:
Set data interval for DagRun triggered from API

uranusjr committed:
Update doc on template variable for AIP-39

uranusjr committed:
Correctly handle None when expanding a DAG ZIP

uranusjr committed:
Bag the DAG for Connexion tests

ashb commented on airflow/api_connexion/endpoints/dag_run_endpoint.py:
```diff
@@ -242,16 +242,23 @@ def post_dag_run(dag_id, session):
+            run_type=DagRunType.MANUAL,
+            data_interval=dag.timetable.infer_data_interval(execution_date),
+            **post_body,
+        )
```
Is there a reason (other than we didn't before) that we don't use `dag.create_run()`?

uranusjr replied:
No; I can change this.

ashb commented on airflow/models/dag.py:
```diff
@@ -112,6 +113,51 @@ def get_last_dagrun(dag_id, session, include_externally_triggered=False):
+    run_after: Optional[pendulum.DateTime]
+
+
+class _NextDagRunInfoCompat(_NextDagRunInfoLegacy):
```
We don't need to maintain compat here (though doing so isn't all that hard) but I think we should give this a better/non-compat-or-private name.

I don't have any brilliant ideas for a name though.

uranusjr replied:
If it’s hard to describe, it’s probably a bad idea 🙂 I’m removing this altogether, we only need to fix a few tests anyway.

ashb commented on airflow/models/dag.py:
```diff
@@ -596,23 +697,17 @@ def get_run_dates(self, start_date, end_date=None, *, align: bool = True):
-        else:
-            start = pendulum.instance(start_date)
+        warnings.warn(
+            "`DAG.get_run_dates()` is deprecated. " "Please use `DAG.iter_dagrun_infos_between()` instead.",
```
```suggestion
            "`DAG.get_run_dates()` is deprecated. Please use `DAG.iter_dagrun_infos_between()` instead.",
```

uranusjr replied:
Black 🤦

ashb commented on airflow/models/dag.py:
```diff
@@ -2372,8 +2471,13 @@ class DagModel(Base):
 
-    # The execution_date of the next dag run
-    next_dagrun = Column(UtcDateTime)
+    # Kept for backwards compatibility. New code should use data_interval columns.
```
I think we will still need this -- it is the execution_date/schedule_date of the new DagRun to create -- which may be different to interval start or end.

```suggestion
```

ashb commented on airflow/models/dag.py:
```diff
@@ -2403,6 +2507,24 @@ def __init__(self, concurrency=None, **kwargs):
+                    "next_dagrun_data_interval_end must be either both None or both datetime"
+                )
+            return None
+        return (self.next_dagrun_data_interval_start, self.next_dagrun_data_interval_end)
```
This seems... odd. What does `dag.data_interval` _mean_? I would expect this at least to be called `next_data_interval` based on the implementation.

uranusjr replied:
Yes, this should be `next_data_interval` instead.

ashb commented on airflow/timetables/interval.py:
```diff
@@ -43,6 +43,13 @@ def __eq__(self, other: Any) -> bool:
+    def infer_data_interval(self, run_after: DateTime) -> Optional[DataInterval]:
+        # Get the last complete period before run_after, e.g. if a DAG run is
+        # scheduled at each midnight, the data interval of a manually trioggered
+        # run at 1am 25th is between 0am 24th and 0am 25th.
```
(Without thinking about this in much detail) I would have expected such a manually triggered run to be from midnight 25th to 1am 25th.

But maybe this way does make more sense. Data intervals aren't (currently) well defined in the case of manual triggers.

uranusjr replied:
What I like about this strategy is we set data interval to a period that always makes sense, instead of some in-between period that doesn’t conform to any DAG scheduling rules. But admittedly I don’t really know what people use manual triggers for (I only use it to check my DAG “can run from start to finish”, for which the data interval doesn’t matter either way since I discard the result anyway) so maybe this is not what they want.

kaxil replied:
Options:

(1) Shouldn't the `data_interval_start_date` be 24th 1 am and `data_interval_end_date` be 25th 1am so `data_interval_end_date` = `schedule_data` / `run_date`. Why the offset of 1 hour?

uranusjr replied:
If the DAG is scheduled for `0 0 * * *`, data intervals always start and end at midnight for all scheduled (and backfilled) runs, so a manual run should match that IMO. If the interval is `timedelta(days=1)` then yes this manual run’s data interval should be between 1am 24th and 1am 25th. I will modify the docstring to clarify this.

ashb commented on airflow/www/views.py:
```diff
@@ -1537,11 +1537,12 @@ def trigger(self, session=None):
             execution_date=execution_date,
-            state=State.QUEUED,
+            data_interval=dag.timetable.infer_data_interval(execution_date),
+            state=State.RUNNING,
```
```suggestion
            state=State.QUEUED,
```

Bad conflict resolve?

uranusjr replied:
Indeed. Thanks for catching this.

ashb reviewed (commented):
I think we should add some tests and docs about the data_interval for when manually triggering runs.

uranusjr committed:
Use DAG.create_dagrun() to create a DagRun

uranusjr committed:
Remove the DagRunInfo compatibility shim

uranusjr committed:
Fix Black blunder

uranusjr committed:
We still need next_dagrun

uranusjr committed:
Rename DagModel.data_interval to next_data_interval

uranusjr committed:
Unify DAG run creation logic in HTML view and API

uranusjr committed:
Remove usage of deprecated get_run_dates()

uranusjr committed:
NamedTuple can't have cached_property

uranusjr committed:
Allow DAGRun to not specify 'conf'

uranusjr committed:
Correctly handle datetime types in tests

uranusjr committed:
Test the CLI triggers DagRun with correct settings

kaxil commented on airflow/models/dag.py:
```diff
@@ -596,23 +649,17 @@ def get_run_dates(self, start_date, end_date=None, *, align: bool = True):
         :return: A list of dates within the interval following the dag's schedule.
```
We should remove align from docstring

https://github.com/apache/airflow/blob/47908878a96775db49eac49339926c0ccff1e3be/airflow/models/dag.py#L644-L648

uranusjr replied:
Thanks for catching this!

kaxil commented on airflow/models/dag.py:
```diff
@@ -579,7 +580,59 @@ def timetable(self) -> Timetable:
+        by their ``schedule_date`` from earliest to latest.
+
+        If ``align`` is ``False``, the first run will happen immediately on
+        ``ealiest``, even if it does not fall on the logical timetable schedule.
```
```suggestion
        ``earliest``, even if it does not fall on the logical timetable schedule.
```

kaxil replied:
Also, an example here with the actual dates might make it easier to understand

kaxil commented on airflow/models/dag.py:
```diff
@@ -579,7 +580,59 @@ def timetable(self) -> Timetable:
+            return
+
+        # If align=False and earliest is not a logical schedule date, "invent"
+        # a data interval betwwen it and the first schedule date.
```
```suggestion
        # a data interval between it and the first schedule date.
```

kaxil commented on airflow/jobs/backfill_job.py:
```diff
@@ -755,8 +759,13 @@ def _execute(self, session=None):
-        run_dates = self.dag.get_run_dates(start_date=start_date, end_date=self.bf_end_date, align=True)
+        # Get DagRun schedule between the start/end dates, which will turn into dag runs.
+        dagrun_start_date = timezone.coerce_datetime(start_date)
+        if self.bf_end_date is None:
+            dagrun_end_date = pendulum.now(timezone.utc)
+        else:
+            dagrun_end_date = pendulum.instance(self.bf_end_date)
```
What do you think about handling this inside `iter_dagrun_infos_between` (similar to what we had earlier in `get_run_dates`) vs not

uranusjr replied:
Personally I dislike the pattern “if end if None, default to now” since it’s too implicit, and users (other contributors of Airflow) can accidentally forget to handle a possibly-None input and cause subtle bugs. For example, say there’s a view to return scheduled run dates of a DAG:

```python
# If the user does not specify, defaults to the DAG’s start_date.
start_date = request.values.get("start_date")

# If the user does not specify, defaults to the DAG’s end_date.
end_date = request.values.get("end_date")

# Oops I forgot to actually check if the user-supplied start_date and end_date is None.

# Now the runs returned are wrong.
runs = list(dag.iter_dagrun_infos_between(start_date, end_date))
```

This can manage to slip through unit tests (if you forgot to do this in code, chances are you wouldn’t remember to test it) and cause either bugs or (worse) incorrect user expectations we need to support for backward compatibility onwards (Hyrum’s law).

Making `iter_dagrun_infos_between` accept only datetimes forces the caller to think it through when exactly the iterator should end, which is a good thing since this can potentially run for a *long* time if `earliest` is a long time in the past.

```python
# Mypy now raises an error here and I’m reminded of the None case.
runs = list(dag.iter_dagrun_infos_between(start_date, end_date))
```

kaxil commented on airflow/migrations/versions/142555e44c17_add_data_interval_start_end_to_dagrun.py:
```diff
@@ -0,0 +1,61 @@
+    if op.get_bind().dialect.name == "mssql":
+        column_type = mssql.DATETIME2(precision=6)
+    else:
+        column_type = UtcDateTime
```
https://github.com/apache/airflow/blob/c384f9b0f509bab704a70380465be18754800a52/airflow/migrations/versions/98271e7606e2_add_scheduling_decision_to_dagrun_and_.py#L48-L55

Do we need something like the above to take care of MySQL too for precision to avoid issues like https://github.com/apache/airflow/pull/9336 . Worth testing this with MySQL

uranusjr replied:
No idea, I’m doing it 🙂

kaxil commented on airflow/timetables/interval.py:
```diff
@@ -43,6 +43,13 @@ def __eq__(self, other: Any) -> bool:
 
+    def infer_data_interval(self, run_after: DateTime) -> Optional[DataInterval]:
+        # Get the last complete period before run_after, e.g. if a DAG run is
+        # scheduled at each midnight, the data interval of a manually trioggered
```
```suggestion
        # scheduled at each midnight, the data interval of a manually triggered
```

kaxil commented on airflow/models/dag.py:
```diff
@@ -2403,6 +2461,24 @@ def __init__(self, concurrency=None, **kwargs):
+                raise AirflowException(
+                    "Inconsistent DagModel: next_dagrun_data_interval_start and "
+                    "next_dagrun_data_interval_end must be either both None or both datetime"
+                )
```
Let's show the values that are passed too, should help in debugging

kaxil commented on airflow/models/taskinstance.py:
```diff
@@ -1530,21 +1532,25 @@ def get_template_context(self, session=None) -> Context:
+        with contextlib.suppress(AttributeError):
+            params.update(task.dag.params)
+
+        dag_run = self.get_dagrun()
```
```suggestion
        dag_run = self.get_dagrun(session=session)
```

kaxil commented on airflow/models/taskinstance.py:
```diff
@@ -1530,21 +1532,25 @@ def get_template_context(self, session=None) -> Context:
+                data_interval_start=None,
+                data_interval_end=None,
+                external_trigger=False,
+                run_id="",
```
```suggestion
                run_id=None,
```

uranusjr replied:
What’s the advantage of None here? IIUC a “real” DagRun cannot have `run_id=None` so I’m using consistent type here in case there’s some function call expecting `str` (of course `run_id` cannot be empty either, but at least the type is correct).

kaxil replied:
No advantage, was just setting it to default value as it's default is None in ctor

uranusjr replied:
Sticking with an empty string for now; we can always change later if None is more advantageous.

kaxil commented on airflow/models/taskinstance.py:
```diff
@@ -1670,37 +1667,97 @@ def get(
+            def deprecated_func():
+                message = (
+                    f"Accessing {key!r} from the template is deprecated and "
+                    f"will be removed in a future version."
```
```suggestion
                    "will be removed in a future version."
```

uranusjr replied:
I like having an extra f here since it aligns the literals :p

kaxil commented on airflow/timetables/base.py:
```diff
@@ -79,6 +74,18 @@ def interval(cls, start: DateTime, end: DateTime) -> "DagRunInfo":
         return cls(run_after=end, data_interval=DataInterval(start, end))
 
+    @property
+    def schedule_date(self) -> DateTime:
```
Shouldn't this be `run_date` based on our proposal?

https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-39+Richer+scheduler_interval#AIP39Richerscheduler_interval-AbstractTimetableinterface

uranusjr replied:
No, this is the “logical” date of the DagRun, not the earliest time to schedule it.

The equivalence of `run_date` in the proposal is `run_after` in the implementation (because I simply forgot `run_date` was in the proposal…); we can change that if needed.

kaxil commented on airflow/models/taskinstance.py:
```diff
@@ -1670,37 +1667,97 @@ def get(
+            return lazy_object_proxy.Proxy(deprecated_func)
+
+        def get_previous_schedule_date() -> Optional[pendulum.DateTime]:
+            ti = self.get_previous_ti(None)
```
```suggestion
            ti = self.get_previous_ti(session=session)
```

kaxil commented on airflow/models/taskinstance.py:
```diff
@@ -1670,37 +1667,97 @@ def get(
+                dt = self.execution_date
+            else:
+                dt = task.dag.previous_schedule(self.execution_date)
+            return timezone.coerce_datetime(dt)
```
Should we suppress deprecation warnings from the following otherwise User will see the deprecation warning but will have no control or can change nothing to change it as the context dict is passed for each operator?

https://github.com/apache/airflow/blob/47908878a96775db49eac49339926c0ccff1e3be/airflow/models/taskinstance.py#L812-L819

uranusjr replied:
This function is used to populate `prev_execution_date`, `prev_ds`, etc., all of them also deprecated. If the user fixes them, this one will go away as well 🙂 

But yeah it’s not a bad idea to suppress this, there’s no need to show two deprecation messages for one thing.

kaxil commented on tests/core/test_core.py:
```diff
@@ -279,23 +281,16 @@ def test_task_get_template(self):
-        assert context['tomorrow_ds'] == '2015-01-02'
-        assert context['tomorrow_ds_nodash'] == '20150102'
+        assert context['data_interval_start'].isoformat() == '2015-01-01T00:00:00+00:00'
+        assert context['data_interval_end'].isoformat() == '2015-01-02T00:00:00+00:00'
```
Let's add test for deprecated configs too

kaxil commented on tests/models/test_dag.py:
```diff
@@ -2015,3 +2016,50 @@ def get_task_instance(session, task):
+        task_id='dummy',
+        dag=dag,
+        owner='airflow',
+    )
```
```suggestion
    DummyOperator(task_id='dummy', dag=dag)
```

for simplicity -- we don't need owner field

uranusjr committed:
Remove unneeded owner field in test DAG

uranusjr committed:
Test deprecated template fields

uranusjr committed:
Pass local session on to child calls

uranusjr committed:
Show wrong values in error messages for debuggability

uranusjr committed:
Typos

uranusjr committed:
Clean up docstring on 'align'

An outdate entry is removed, and an example to explain the behavior.

uranusjr committed:
Improve datetime field handling in migration

uranusjr committed:
Fix template context type tests

uranusjr committed:
Create an actual DagRun to populate data interval

uranusjr committed:
Test to ensure /trigger populated data interval

uranusjr committed:
Include dagrun in the data interval property name

uranusjr committed:
Test data intervals around in the scheduler job

uranusjr committed:
Rename schedule_date to logical_date

Waiting for updates in the dev mailing list

https://lists.apache.org/thread.html/r00d88d22ae15f9c98b91c8f28163c09e59fb4a6436586861c69a66b7%40%3Cdev.airflow.apache.org%3E

ephraimbuddy commented on airflow/api_connexion/endpoints/dag_run_endpoint.py:
```diff
@@ -242,21 +243,29 @@ def post_dag_run(dag_id, session):
+            conf=post_body.get("conf"),
+            external_trigger=True,
+            dag_hash=current_app.dag_bag.dags_hash.get(dag_id),
+        )
```
In line 239 above, we queried the DagModel to check if dag exists. I'm thinking that instead of `current_app.dag_bag.get_dag(dag_id).create_dagrun` we should get dag from line 239:
```dag = session.query(DagModel).filter(DagModel.dag_id == dag_id).first()``` and then `dag.create_dagrun`, what do you think?

uranusjr replied:
`current_app.dag_bag.get_dag(dag_id)` returns a `DAG`, not a `DagModel`. only `DAG` can call `create_dagrun` (yes it’s a bit confusing, it took a while for me quite a while to sort these out).

The new implementation matches more closely to how the Trigger Run web UI is doing.

https://github.com/apache/airflow/blob/ff75cbcac9ec0b1992b4fddd6c160901f23e0c2a/airflow/www/views.py#L1570-L1621

ephraimbuddy replied:
Oh, nice

uranusjr committed:
Revert www/views.py change

uranusjr committed:
Make a DagRun always have a data interval

uranusjr committed:
Make data interval 'exact' instead of None

uranusjr committed:
Back template variables against data interval

This makes logical date (i.e. execution data) only an internal concept
and never directly exposed to users. Template context variables now are
always backed by data_interval_[start|end], and variables that don't
make sense against it are deprecated.

This introduces one slightly backward-incompatible change. Since ds
and its 'nodash' variants are now backed by data_interval_start,
their meanings change slightly for manually-triggered DAG runs -- the
logical date of such a run is after the *end* of the data interval, so
the value of ds and ds_nodash is shifted one interval to the past. But
this actually makes more sense, so we're treating this as a bug fix.

uranusjr committed:
Add data interval to tests

Since we changed the meaning of {{ ds }}, template tests using it needs
some modification to continue working. Specifically, they need to
specify data_interval so to not receive an incorrect inferred interval.

ashb commented on airflow/www/views.py:
```diff
@@ -1602,7 +1602,7 @@ def trigger(self, session=None):
                 )
 
-        dag = current_app.dag_bag.get_dag(dag_id)
+        dag: DAG = current_app.dag_bag.get_dag(dag_id)
```
```suggestion
        dag = current_app.dag_bag.get_dag(dag_id)
```

Revert this as we don't otherwise touch this file.

ashb commented on docs/apache-airflow/macros-ref.rst:
```diff
@@ -83,6 +78,27 @@ Variable                                Description
+``{{ prev_execution_date }}``           the previous execution date (if available) (`pendulum.Pendulum`_)
+``{{ prev_execution_date_success }}``   execution date from prior successful dag run,
+                                        same as ``prev_logical_date_success``
+``{{ next_execution_date }}``           the next execution date, same as ``next_logical_date``
```
```suggestion
``{{ next_execution_date }}``           the next execution date, same as ``data_interval_end``
```

(Yes, it would be the same, but I think we should point people to use the explicit data interval properties)

ashb commented on docs/apache-airflow/macros-ref.rst:
```diff
@@ -35,28 +35,23 @@ in all templates
+``{{ next_ds }}``                       the next logical date as ``YYYY-MM-DD``
+                                        if ``{{ ds }}`` is ``2018-01-01`` and ``schedule_interval`` is ``@weekly``,
+                                        ``{{ next_ds }}`` will be ``2018-01-08``
+``{{ next_ds_nodash }}``                the next schedule date as ``YYYYMMDD`` if exists, else ``None``
```
```suggestion
``{{ next_ds_nodash }}``                the next logical date as ``YYYYMMDD`` if exists, else ``None``
```

ashb commented on docs/apache-airflow/macros-ref.rst:
```diff
@@ -35,28 +35,23 @@ in all templates
-``{{ tomorrow_ds_nodash }}``            the day after the execution date as ``YYYYMMDD``
-``{{ ts }}``                            same as ``execution_date.isoformat()``. Example: ``2018-01-01T00:00:00+00:00``
+``{{ logical_date }}``                  the logical date of the DAG run (`pendulum.Pendulum`_)
+``{{ ds }}``                            the logical date as ``YYYY-MM-DD``
```
I think this should be derrived from data_interval_start, rather than logical date.

Previously the implicit data interval was exection_date...next_exectuion_date

uranusjr replied:
How about `@once` DAGs? We decided to set its data interval to None, should this also become None?

ashb replied:
Hmmmmmm, _Probably_, but would make it breaking change :(

ashb replied:
Maybe we need to leave this as it is but deprecate it, and replace with `{{ data_interval_start | ds }}` (i.e. add new `ds`, `ds_nodash`, `ts`, etc filters) -- separate PR there.

ashb commented on docs/apache-airflow/macros-ref.rst:
```diff
@@ -35,28 +35,23 @@ in all templates
+``{{ logical_date }}``                  the logical date of the DAG run (`pendulum.Pendulum`_)
+``{{ ds }}``                            the logical date as ``YYYY-MM-DD``
+``{{ ds_nodash }}``                     the logical date as ``YYYYMMDD``
+``{{ ts }}``                            same as ``schedule_date.isoformat()``. Example: ``2018-01-01T00:00:00+00:00``
```
```suggestion
``{{ ts }}``                            same as ``logical_date.isoformat()``. Example: ``2018-01-01T00:00:00+00:00``
```

But see previous comment about deriving form data interval.

ashb commented on airflow/models/dag.py:
```diff
@@ -537,30 +540,31 @@ def next_dagrun_info(
             but not manual).
-        :return: A 2-tuple containing the DagRun's execution date, and the
-            earliest it could be scheduled.
+        :returns: DagRunInfo of the next dagrun, or None if a dagrun is not
```
`:returns:` vs `:return:
```suggestion
        :return: DagRunInfo of the next dagrun, or None if a dagrun is not
```
I think?

uranusjr replied:
Both are used in the code base so I think Sphinx is fine with either. But this is indeed unintended 🙂

uranusjr committed:
More consistently use :return:

Co-authored-by: Ash Berlin-Taylor <ash_github@firemirror.com>

uranusjr committed:
Point people to the template filters
